crates/maple_cli/src/cmd/tagfiles.rs

```diff
@@ -0,0 +1,260 @@
+}
+
+impl TagInfo {
+    pub fn format(&self, cwd: &PathBuf, winwidth: usize) -> String {
+        let name = format!("{} ", self.name);
+        let taken_width = name.len() + 1;
+        let path_len = self.path.len() + 2;
+        let mut adjustment = 0;
+
+        let mut home_path = PathBuf::new();
+        let path = Path::new(&self.path);
+        let path = path.strip_prefix(cwd).unwrap_or(
+            // FIXME: is there a way to avoid cloning HOME?
+            HOME.clone()
+                .map(|home| {
+                    path.strip_prefix(home)
+                        .map(|path| {
+                            home_path.push("~");
+                            home_path = home_path.join(path);
+                            home_path.as_path()
+                        })
+                        .unwrap_or(path)
+                })
+                .unwrap_or(path),
+        );
+        let path = path.display().to_string();
+
+        let path_label = if taken_width > winwidth {
+            format!("[{}]", path)
+        } else {
+            let available_width = winwidth - taken_width;
+            if path_len > available_width && available_width > 3 {
+                let diff = path_len - available_width;
+                adjustment = 2;
+                format!("[…{}]", path.chars().skip(diff + 2).collect::<String>())
+            } else {
+                format!("[{}]", path)
+            }
+        };
+
+        let path_len = path_label.len();
+        let text_width = if path_len < winwidth {
+            winwidth - path_len
+        } else {
+            winwidth
+        } + adjustment;
+
+        format!(
+            "{text:<text_width$}{path_label}:::{path}:::{pattern}",
+            text = name,
+            text_width = text_width,
+            path_label = path_label,
+            path = self.path,
+            pattern = self.pattern,
+        )
+    }
+
+    pub fn parse(base: &PathBuf, input: &str) -> Result<Self, std::string::ParseError> {
+        let mut field = 0;
+        let mut index = 0;
+        let mut index_last = 0;
+
+        let mut name = String::from("");
+        let mut path = String::from("");
+        let mut pattern = String::from("");
+        let mut kind = String::from("");
+
+        for c in input.chars() {
+            if c == '\t' {
+                match field {
+                    0 => name = String::from(&input[index_last..index]),
+                    1 => {
+                        let path_buf = base.join(String::from(&input[index_last..index]));
+                        path = path_buf.as_path().display().to_string();
+                    }
+                    2 => pattern = String::from(&input[index_last..index]),
+                    3 => kind = String::from(&input[index_last..index]),
+                    _ => {}
+                }
+                field += 1;
+                index_last = index + c.len_utf8();
+            }
+            index += c.len_utf8();
+        }
```

Comment:
Just use `char_indices`.

Reply:
Also, you allocate a new string each time you call a `String::from`. It is better to call `name.clear(); name.push_str(…);`.

Reply:
For `path` it would be

```rust
use std::fmt::Write;
path.clear();
write!(&mut path, "{}", path_buf.display()).unwrap();
```

Reply:
Got it. For this particular case, the `.clear()` isn't necessary right?